aten/src/ATen/native/hammerblade/Offload.cpp

```diff
@@ -85,11 +82,97 @@ static void cleanup_device(std::vector<eva_t> args, std::vector<eva_t> ptrs) {
  * Offloading wrapper
  * --------------------------------------------------------------------------*/
 
-void offload_op_binary_impl(TensorIterator& iter, Scalar alpha, 
-    const char* kernel) {
+//=======================================================================
+// Offloading operations that have tensors and scalars as arguments
+//=======================================================================
+
+void offload_tensor_scalar_impl(std::vector<Tensor> tensors, std::vector<Scalar> scalars,
+                                const char* kernel) {
+
+  // Device pointers to tensors on the device
+  std::vector<eva_t> device_args;
+  std::vector<eva_t> device_ptrs;
+
+  // Allocate device tensors and copy the data
+  for(int i=0; i<tensors.size(); i++) {
+    // Iterate over all tensors to create
+    // corresponding tensors on the device.
+    auto arg = tensors[i];
+    TORCH_INTERNAL_ASSERT(arg.device().is_hammerblade())
+    // Read low-level meta-data from argument tensor
+    uint64_t n = arg.numel();
+    uint32_t dims = arg.dim();
+    const int64_t* strides = arg.strides().data();
+    TORCH_INTERNAL_ASSERT(arg.strides().size() == dims);
+    const void* data = arg.data_ptr();
+    // Create raw-tensor
+    eva_t device_arg = create_device_tensor(n, dims,
+        strides, data, device_ptrs);
+    device_args.push_back(device_arg);
+    // NOTE: here we are assuming all strides need to be copied.
+  }
+
+  // Allocate device scalars and copy the data
+  for(int i=0; i<scalars.size(); i++) {
+    auto alpha = scalars[i];
+    device_args.push_back(create_device_scalar(alpha.to<float>()));
+  }
+
+  c10::hammerblade::offload_kernel(kernel, device_args);
+
+  // Need to deallocate those args on device
+  cleanup_device(device_args, device_ptrs);
+
+}
+
+//=======================================================================
+// Offloading operations that use TensorIterator
+//=======================================================================
+
+void offload_iterator_op_impl(TensorIterator& iter, Scalar alpha,
+    const char* kernel, uint32_t ntensors) {
 
   TORCH_INTERNAL_ASSERT(iter.can_use_32bit_indexing());
-  TORCH_INTERNAL_ASSERT(iter.ntensors() == 3); // output, input1, and input2
+  TORCH_INTERNAL_ASSERT(iter.ntensors() == ntensors);
+
+  // It is very important to use serial_for_each here, since we assume a single
+  // HammerBlade device in the system
+  iter.serial_for_each([&](char** data, const int64_t* strides, int64_t n) {
```

Comment:
I believe it's okay to use just `for_each` here, because it doesn't assume multiple devices but only multiple host threads. And, we disable host multi-threading by the repo level parameter `USE_OPENMP=0` as `libbsg_manycore_runtime.so` is not thread-safe yet.

Reply:
I was playing safe ;) since I noticed that pytorch still links against pthreads